File: WINGS_AI_Bootcamp/Haskell/largestPrimeFactor.hs
largestPrimeFactor :: Integer -> Integer
largestPrimeFactor n = lpf n 2 (-1)
  where
    lpf n f largest
      | n < 2            = largest
      | n `mod` f == 0   = lpf (n `div` f) f f
      | f * f > n        = if n > 1 then n else largest
      | otherwise        = lpf n (nextFactor f) largest

    nextFactor 2 = 3
    nextFactor x = x + 2

main :: IO ()
main = print $ largestPrimeFactor 315  -- Output: 7

File: WINGS_AI_Bootcamp/Elixir/largestPrimeFactor.ex
defmodule Factors do
  def largest_prime_factor(n), do: lpf(n, 2, -1)

  defp lpf(n, _f, largest) when n < 2, do: largest

  defp lpf(n, f, _largest) when rem(n, f) == 0 do
    lpf(div(n, f), f, f)
  end

  defp lpf(n, f, largest) when f * f > n do
    if n > 1, do: n, else: largest
  end

  defp lpf(n, f, largest) do
    lpf(n, next_factor(f), largest)
  end

  defp next_factor(2), do: 3
  defp next_factor(x), do: x + 2
end

IO.inspect Factors.largest_prime_factor(315)  # Output: 7

